pubsub/pubsub.go

```diff
@@ -562,20 +562,37 @@ func (s *Subscription) Receive(ctx context.Context) (_ *Message, err error) {
 			// log.Printf("BATCH SIZE %d", batchSize)
 
 			go func() {
+				defer func() {
+					close(s.waitc)
+					s.waitc = nil
+				}()
 				if s.preReceiveBatchHook != nil {
 					s.preReceiveBatchHook(batchSize)
 				}
-				msgs, err := s.getNextBatch(batchSize)
-				s.mu.Lock()
-				defer s.mu.Unlock()
-				if err != nil {
-					// Non-retryable error from ReceiveBatch -> permanent error.
-					s.err = err
-				} else if len(msgs) > 0 {
-					s.q = append(s.q, msgs...)
+				resultChannel := s.getNextBatch(batchSize)
+				for {
+					select {
+					case msgs, ok := <-resultChannel.msgs:
+						if !ok {
+							// batch reception finished
+							return
+						} else if len(msgs) > 0 {
+							// messages received from channel
+							s.mu.Lock()
+							s.q = append(s.q, msgs...)
+							s.mu.Unlock()
+						}
+					case err := <-resultChannel.err:
+						// err can receive message only after batch group completes
+						if err != nil {
+							// Non-retryable error from ReceiveBatch -> permanent error.
+							s.mu.Lock()
+							s.err = err
+							s.mu.Unlock()
+						}
+						return
```

Comment:
I think you can drop this `return` to make the flow simpler; the next time through the `select` should get `ok=false` and exit there.